client/src/main/java/org/apache/celeborn/client/ShuffleClientImpl.java

```diff
@@ -1586,4 +1708,134 @@ private boolean connectFail(String message) {
+                        req.reviveStatus = StatusCode.SUCCESS;
+                      } else if (mapperEnded(shuffleId, req.mapId)) {
+                        req.reviveStatus = StatusCode.MAP_ENDED;
+                      } else {
+                        filteredRequests.add(req);
+                        mapIds.add(req.mapId);
+                        PartitionLocation loc = req.loc;
+                        if (!partitionMaxEpoch.containsKey(loc.getId())
+                            || partitionMaxEpoch.get(loc.getId()) < loc.getEpoch()) {
+                          requestsToSend.put(loc.getId(), req);
+                          partitionMaxEpoch.put(loc.getId(), loc.getEpoch());
+                        }
+                      }
+                    }
+
+                    if (!requestsToSend.isEmpty()) {
+                      int length = requestsToSend.size();
+                      int[] partitionIds = new int[length];
+                      int[] epochs = new int[length];
+                      StatusCode[] causes = new StatusCode[length];
+                      PartitionLocation[] locs = new PartitionLocation[length];
+                      int idx = 0;
+                      for (ReviveRequest req : requestsToSend.values()) {
+                        partitionIds[idx] = req.partitionId;
+                        epochs[idx] = req.loc.getEpoch();
+                        causes[idx] = req.cause;
+                        locs[idx] = req.loc;
+                        idx++;
+                      }
+
+                      // Call reviveBatch. Return null means Exception caught
+                      Map<Integer, StatusCode> results =
+                          reviveBatch(appId, shuffleId, mapIds, partitionIds, epochs, locs, causes);
+                      if (results == null) {
+                        for (ReviveRequest req : filteredRequests) {
+                          req.reviveStatus = StatusCode.REVIVE_FAILED;
+                        }
+                      } else {
+                        for (ReviveRequest req : filteredRequests) {
+                          if (mapperEnded(shuffleId, req.mapId)) {
+                            req.reviveStatus = StatusCode.MAP_ENDED;
+                          } else {
+                            req.reviveStatus = results.get(req.partitionId);
+                          }
+                        }
+                      }
+                    }
+                  } // End foreach shuffleMap.entrySet
+                } // End foreach inProcessRequests.entrySet
+              }
+
+              inProcessRequests.clear();
+            }
+          },
+          100,
+          100,
+          TimeUnit.MILLISECONDS);
+    }
+
+    public void addRequest(String appId, int shuffleId, ReviveRequest request) {
+      excludeWorkerByCause(request.cause, request.loc);
+      // This sync is necessary to ensure the add action is atomic
+      synchronized (this) {
```

Comment:
This synchronization block can be replaced with ReentrantReadWriteLock's ReadLock. Adding requests should not block others from adding requests.

Reply:
> 

Actually I thought about it but after I saw this blog I decided not to use it. https://www.javacodegeeks.com/2014/06/java-8-stampedlocks-vs-readwritelocks-and-synchronized.html

5 readers vs. 5 writers:
![image](https://github.com/apache/incubator-celeborn/assets/948245/23251f6a-680e-492f-ae16-f770280192e2)
10 readers vs. 10 writers:
![image](https://github.com/apache/incubator-celeborn/assets/948245/d1fd7016-0d8c-4f57-b3e1-64e26507d254)
16 readers vs. 4 writers:
![image](https://github.com/apache/incubator-celeborn/assets/948245/1ce0cfa8-1a8b-44bc-88ce-a29761e35c27)
19 readers vs. 1 writer:
![image](https://github.com/apache/incubator-celeborn/assets/948245/3535084b-aeb8-4004-9149-3be298d3ba2b)